### DISCUSSION TITLE
On refresh the screen: "The provided `href` value is missing query values to be interpolated properly."

### DISCUSSION THREAD

**Discussion by guscsales** [vercel/next.js#33657] [Help] [answered]
### Run `next info` (available from version 12.0.8 and up)

```
Operating System:
      Platform: darwin
      Arch: arm64
      Version: Darwin Kernel Version 21.1.0: Wed Oct 13 17:33:24 PDT 2021; root:xnu-8019.41.5~1/RELEASE_ARM64_T8101
    Binaries:
      Node: 16.13.0
      npm: 8.1.0
      Yarn: 1.22.17
      pnpm: N/A
    Relevant packages:
      next: 12.0.8
      react: 17.0.2
      react-dom: 17.0.2
```

### What version of Next.js are you using?

12.0.8

### What version of Node.js are you using?

16.13.0

### What browser are you using?

Chrome

### What operating system are you using?

macOS

### How are you deploying your application?

Vercel

### Describe the Bug

This error does not in fact causes an issue on the screen, but running the server locally when I refresh a dynamic route with a custom param the application returns the following error:

```
Unhandled Runtime Error
Error: The provided `href` (/incidents/[incidentId]/triage) value is missing query values (incidentId) to be interpolated properly. Read more: https://nextjs.org/docs/messages/href-interpolation-failed
```

This happens only when refreshing the dynamic route, when I use the `Link` or `router.push` to navigate this does not happen.

My path to this page is: `/pages/incidents/[incidentId]/index.tsx` (using on the browser `http://localhost:4200/incidents/0f421737-d616-46a8-8bd9-8871e92de716/triage`) and also I'm using `nx@13.5.8`.

A gif to explain more. Note when I go to the previous page and click on the card nothing happens, but when refreshing the page the error happens.

![Kapture 2022-01-25 at 15 28 58](https://user-images.githubusercontent.com/5130296/151037330-2d7d8099-056f-48a3-9c4e-721d8afc59f7.gif)

### Expected Behavior

Render the screen after a refresh the browser without this error.

### To Reproduce

Create a dynamic screen using a path like this: `/pages/incidents/[incidentId]/triage/index.tsx`, then enter a URL like: `http://localhost:4200/incidents/0f421737-d616-46a8-8bd9-8871e92de716/triage`.

**Comment by wwwdepot** [comment]
Error is pretty much straightforward, your query `incidentId` is undefined. How do you get `incidentId`, are you using useRouter or getStaticProps?

**Reply by guscsales** [reply]
I was using the `useRouter` - and it's possible to retrieve the value, I can get the uuid on the URL without issues - but even commenting the hook the error persists. So without getting the `incidentId` the errors continue happening, for this I felt this was strange.

**Reply by wwwdepot** [reply]
Your worst case scenario is to go through each sub-component that uses `incidentId`.

**Reply by JavierMartinz** [reply]
I'm also seeing tons of errors like these in Sentry since a few days ago, but I couldn't reproduce it locally nor in production. I saw your discussion and I thought I could reproduce it but it's not happening for me the way it is for you... I don't know what else to test or to debug in order to fix it.

**Comment by guscsales** [comment]
I've executed the getStaticPaths to get the values, and I've gotten the same error.

Using the `useRouter` the first log returned `undefined` and the second one returned the params correctly. The image below is the console.log for `const {params} = router.query`
![image](https://user-images.githubusercontent.com/5130296/151167878-a7555db1-dff3-4291-9a3b-425557399f55.png)


This is from `getStaticPaths` (I just change the parameters a little, I'm using the [...params], but the error is the same).
```
export async function getStaticProps(context) {
  console.log(context);
  return {
    props: {
      foo: false,
    },
  };
}

export async function getStaticPaths() {
  return {
    paths: [],
    fallback: true,
  };
}
```
<img width="683" alt="image" src="https://user-images.githubusercontent.com/5130296/151172168-92df7d61-4801-4640-9e71-8586fe25cbf9.png">


Getting from `getServerSideProps` there is no undefined, the params are returned well, but the error on the screen persists.
```
export async function getServerSideProps(context) {
  const {
    params: [incidentId, incidentDetailScreen],
  } = context.query;

  return { props: { incidentId, incidentDetailScreen } };
}
```

Both is working, but the error persists on the screen.

**Reply by wwwdepot** [reply]
Is the bug reproducible in a new project? I can't reproduce it in a blank new next.js app, tried ssr/static pages, everything is working fine on my machine Win10 x64, also tried in ubuntu vm.

- Is this new in v12.0.8?
- Call Stack/error logs in console, does it help narrowing down to which component that error occurs in?
- Does the problem occurs in other pages, it could something in the `_app.js`
- Try disable swc by using a custom `.babelrc`, [see Docs](https://nextjs.org/docs/advanced-features/customizing-babel-config).
- If the error only occurs in `triage` page: go to the page, comment out each sub-component and reload page to find out which sub-component is causing the problem, go from top down `page > subcomponent > sub-subcomponent`

**Reply by philippe-solosegment** [reply]
I'm also experiencing the same issue, was there any movement on this issue?

**Comment by guscsales** [comment, marked as answer]
In fact, we found the error here, it was an internal error. We were doing a redirect without spreading the whole URL, only the pathname, and because of this, the error was happening.

Before: 
```
if (user?.id) {
      replace(pageToRedirect);
    }
```

Now:
```
if (user?.id) {
replace({ pathname: pageToRedirect, query });
    }
```

This fixed the error for me. Thanks for all help!

**Reply by sinclairnick** [reply]
@ayasalam0
Consider the route `plans/[id]`
If `pageToRedirect = "/plans` then `id` is never supplied, and must be supplied via the query object:

`router.replace(pageToRedirect, { id: 123})`

Equivalently, you could do `router.replace(pageToRedirect + "/123")`

**Reply by volkmen** [reply]
query param MUST be  there `router.replace({ pathname: pageToRedirect, query })` to avoid such issue.
Thank you @guscsales

**Comment by maura-dev** [comment]
> In fact, we found the error here, it was an internal error. We were doing a redirect without spreading the whole URL, only the pathname, and because of this, the error was happening.
> 
> 
> 
> Before: 
> 
> ```
> 
> if (user?.id) {
> 
>       replace(pageToRedirect);
> 
>     }
> 
> ```
> 
> 
> 
> Now:
> 
> ```
> 
> if (user?.id) {
> 
> replace({ pathname: pageToRedirect, query });
> 
>     }
> 
> ```
> 
> 
> 
> This fixed the error for me. Thanks for all help!

Hello please can you help explain in detail how this error was fixed

**Comment by YeudaBy** [comment]
I have a similar problem, in the first render of the page, the error appears, but after a fast refresh it disappears. For the test I entered the following code:

```
const Id = (props) => {
    return null
}

export default Id; 
```

and still.
it's the only code i have in this file, `/pages/tests/[id].js`.
this append to me even if i opened a new tab and entered the url without any navigation.

**Comment by YeudaBy** [comment]
Yes, the problem was in the layout layer, at the BottomNavigaion, if I
remember correctly.
I used router.push some dynamic ids, so I guess that causes the problem. I
replaced it with hard coded values and the problem solved.

**Comment by zeelmodi-simform** [comment]
@guscsales Thanks a lot. I found the real culprit for getting this error. There was `router.replace(router.pathname)` written in my code.
